src/data/polynomial/lifts.lean

```diff
@@ -0,0 +1,332 @@
+  lifts_deg f p :=
+begin
+  have h : p.support.card = 0 ∨ p.support.card = 1 := by omega,
+  cases h with h0 h1,
+  { simp only [finsupp.support_eq_empty, finset.card_eq_zero] at h0,
+    use 0,
+    simp only [h0, degree_zero, eq_self_iff_true, and_self, map_zero] },
+  have hzero : p ≠ 0 := by rw [←nonempty_support_iff, ←finset.card_pos, h1]; exact zero_lt_one,
+  replace hs := C_mul_X_pow_eq_self (le_of_eq h1),
+  rw [← hs, lifts_deg],
+  obtain ⟨q, hq⟩ := hl,
+  use (C(q.coeff p.nat_degree) * X ^ p.nat_degree),
+  have hcoeff : f (q.coeff p.nat_degree ) = p.coeff p.nat_degree,
+  { rw [← hq, coeff_map] },
+  split,
+  { simp only [leading_coeff, hcoeff, map_C, map_pow, map_X, map_mul] },
+  have pcoeff : p.leading_coeff ≠ 0 := by intro ha; exact hzero (leading_coeff_eq_zero.1 ha),
+  have qcoeff : q.coeff p.nat_degree ≠ 0,
+  { intro habs,
+    rw [habs, ←leading_coeff, ring_hom.map_zero] at hcoeff,
+    exact pcoeff hcoeff.symm },
+  simp only [qcoeff, pcoeff, ne.def, not_false_iff, degree_monomial]
+end
+
+/--A polynomial lifts if and only if it can be lifted to a polynomial of the same degree. -/
+lemma lifts_deg_iff_lifts (f : R →+* S) (p : polynomial S) : lifts_deg f p ↔ lifts f p :=
+begin
+  split,
+  { intro h; obtain ⟨q, hq⟩ := h; use q; exact hq.1 },
+  generalize' hd : p.nat_degree = d,
+  revert hd p,
+  apply nat.strong_induction_on d,
+  intros n hn p hdeg hlifts,
+  by_cases hsuppcard : 2 ≤ p.support.card,
+  swap,
+  { rw [not_le] at hsuppcard; exact lifts_deg_of_supp_le_one hsuppcard hlifts },
+  obtain ⟨lead, hlead⟩ := lifts_deg_of_supp_le_one (lt_of_le_of_lt
+    (@card_support_C_mul_X_pow_le_one S _ p.leading_coeff p.nat_degree) one_lt_two)
+    (lead_monom_lifts_of_lifts hlifts),
+  have pzero : p ≠ 0,
+  { rw [← nonempty_support_iff, ← finset.card_pos]; exact lt_of_lt_of_le zero_lt_two hsuppcard },
+  have lead_zero : p.leading_coeff ≠ 0,
+  { rw [ne.def, leading_coeff_eq_zero]; exact pzero },
+  have deg_lead : lead.degree = p.nat_degree,
+  { rw [hlead.2, degree_monomial p.nat_degree lead_zero] },
+  have deg_erase := erase_lead_nat_degree_lt hsuppcard,
+  rw hdeg at deg_erase,
+  obtain ⟨erase, herase⟩ := hn p.erase_lead.nat_degree deg_erase p.erase_lead
+    (refl p.erase_lead.nat_degree) (erase_lead_lifts_of_lifts hlifts),
+  use erase + lead,
+  split,
+  { simp only [hlead, herase, map_add]; nth_rewrite 3 ← erase_lead_add_C_mul_X_pow p },
+  rw [←hdeg] at deg_erase,
+  replace deg_erase := lt_of_le_of_lt degree_le_nat_degree (with_bot.coe_lt_coe.2 deg_erase),
+  rw [← deg_lead, ← herase.2] at deg_erase,
+  rw [degree_add_eq_of_degree_lt deg_erase, deg_lead, degree_eq_nat_degree pzero]
+end
+
+end lift_deg
+
+section monic
+
+/--`lifts_deg_monic f p` means that `p` can be lifted to a monic polynomial of the same degree. -/
+def lifts_deg_monic (f : R →+* S) (p : polynomial S) : Prop :=
+  ∃ (q : polynomial R), map f q = p ∧ q.degree = p.degree ∧ q.monic
```

Comment:
Again, only a monic polynomial can lift to a monic polynomial, so a lemma `polynomial.lifts.exists_monic` is what I'd add here instead of a new `def`.

Reply:
Thank you for all your comments. I will update the source file tomorrow.

Reply:
I've updated the code taking into account your comments.